### PR TITLE
[#10650] Add null check for getFeedbackSession in API actions

### LINKED ISSUE #10650: NPE in verifySessionOpenExceptForModeration if session is null
Received a stream of these today.  Not sure if it is a one-off corner case of something systemically wrong with the check. Have a look at the code first. If there is no systemic problem, I can try to find more info about the specific case.

>Error Message: teammates.common.util.Logger severe: NullPointerException caught by WebApiServlet:
java.lang.NullPointerException
at teammates.ui.webapi.action.BasicFeedbackSubmissionAction.verifySessionOpenExceptForModeration(BasicFeedbackSubmissionAction.java:152)
at teammates.ui.webapi.action.ConfirmFeedbackSessionSubmissionAction.checkSpecificAccessControl(ConfirmFeedbackSessionSubmissionAction.java:33)
at teammates.ui.webapi.action.Action.checkAccessControl(Action.java:94)
at teammates.ui.webapi.action.WebApiServlet.invokeServlet(WebApiServlet.java:73)
at teammates.ui.webapi.action.WebApiServlet.doPost(WebApiServlet.java:47)
at javax.servlet.http.HttpServlet.service(HttpServlet.java:707)
at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:848)
at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1772)
at teammates.ui.webapi.action.OriginCheckFilter.doFilter(OriginCheckFilter.java:98)
at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1759)
at com.googlecode.objectify.ObjectifyFilter.doFilter(ObjectifyFilter.java:48)
at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1759)
at com.google.apphosting.utils.servlet.JdbcMySqlConnectionCleanupFilter.doFilter(JdbcMySqlConnectionCleanupFilter.java:60)
at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1759)
at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:582)
at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:524)
at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:226)
at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:134)
at com.google.apphosting.runtime.jetty9.ParseBlobUploadHandler.handle(ParseBlobUploadHandler.java:119)
at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1182)
at com.google.apphosting.runtime.jetty9.AppEngineWebAppContext.doHandle(AppEngineWebAppContext.java:187)
at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:512)
at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1112)
at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
at com.google.apphosting.runtime.jetty9.AppVersionHandlerMap.handle(AppVersionHandlerMap.java:293)
at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:134)
at org.eclipse.jetty.server.Server.handle(Server.java:539)
at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:333)
at com.google.apphosting.runtime.jetty9.RpcConnection.handle(RpcConnection.java:213)
at com.google.apphosting.runtime.jetty9.RpcConnector.serviceRequest(RpcConnector.java:81)
at com.google.apphosting.runtime.jetty9.JettyServletEngineAdapter.serviceRequest(JettyServletEngineAdapter.java:134)
at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.dispatchServletRequest(JavaRuntime.java:761)
at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.dispatchRequest(JavaRuntime.java:724)
at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.run(JavaRuntime.java:694)
at com.google.apphosting.runtime.JavaRuntime$NullSandboxRequestRunnable.run(JavaRuntime.java:880)
at com.google.apphosting.runtime.ThreadGroupPool$PoolEntry.run(ThreadGroupPool.java:270)
at java.lang.Thread.run(Thread.java:748)

### PR DESCRIPTION
**Pull request by madanalogy**
Fixes #10650

### CODE REVIEW

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]
This is not right. At the point this method is reached, we should already be ensured that `feedbackSession` is not null (otherwise `null` can be other reasons other than not found). The check for not found should be done early at the `execute` and possibly `checkSpecificAccessControl` block.

**Comment by madanalogy** [comment]
> This is not right. At the point this method is reached, we should already be ensured that `feedbackSession` is not null (otherwise `null` can be other reasons other than not found). The check for not found should be done early at the `execute` and possibly `checkSpecificAccessControl` block.

Added exception to be thrown from the DB layer. If you're okay with the implementation I'll go fix the test cases

File: src/main/java/teammates/storage/api/FeedbackSessionsDb.java [hunk 1]
```diff
@@ -80,7 +81,15 @@ public FeedbackSessionAttributes getFeedbackSession(String courseId, String feed
...
-            return null;
+        }
+        if (feedbackSession == null || feedbackSession.isSessionDeleted()) {
+            String errorMessage = "";
+            if (feedbackSession == null) {
+                errorMessage = "Session not found.";
+            } else {
+                errorMessage = "Session is deleted.";
+            }
+            throw new EntityNotFoundException(new EntityDoesNotExistException(errorMessage));
```
Comment (xpdavid) [inline src/main/java/teammates/storage/api/FeedbackSessionsDb.java:92]: This violates our policy on DB layer though https://github.com/TEAMMATES/teammates/blob/master/docs/design.md#policies-2

Comment (wkurniawan07) [reply src/main/java/teammates/storage/api/FeedbackSessionsDb.java:92]: There are two very dangerous things done here:
- Assuming that feedback session not found is not desirable. DB layer has no concern with such.
- Throwing unchecked exception in DB layer.

Comment (madanalogy) [reply src/main/java/teammates/storage/api/FeedbackSessionsDb.java]: The problem was that this issues is quite prevalent. There is almost never a null check for `logic.getFeedbackSession()` in the access control of feedback session actions. I've moved the check to `checkSpecificAccessControl`

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]

File: src/main/java/teammates/ui/webapi/action/ConfirmFeedbackSessionSubmissionAction.java [hunk 2]
```diff
@@ -30,6 +32,10 @@ public void checkSpecificAccessControl() {
         String feedbackSessionName = getNonNullRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_NAME);
         FeedbackSessionAttributes feedbackSession = logic.getFeedbackSession(feedbackSessionName, courseId);
 
+        if (feedbackSession == null) {
+            throw new EntityNotFoundException(new EntityDoesNotExistException("Feedback session not found"));
+        }
```
[the remark this answers is not in the document]
Comment (madanalogy) [reply src/main/java/teammates/ui/webapi/action/ConfirmFeedbackSessionSubmissionAction.java:37]: That's going to take a while. What if I just move the validation to the logic layer? i.e. throw exception if not found at `logic.getFeedbackSession()`

[the remark this answers is not in the document]
Comment (wkurniawan07) [reply src/main/java/teammates/ui/webapi/action/ConfirmFeedbackSessionSubmissionAction.java:37]: - Logic has no business with throwing that kind of exception
- Even if it does, logic layer must not throw any runtime exception

File: src/main/java/teammates/ui/webapi/action/Action.java [hunk 3]
```diff
@@ -190,6 +193,14 @@ protected String getRequestBody() {
         return requestBody;
     }
 
+    protected FeedbackSessionAttributes getFeedbackSession(String feedbackSessionName, String courseId) {
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/webapi/action/Action.java:196]: Just `getFeedbackSession` doesn't immediately tell that it's going to throw some exception if the value is null. Name it something like `getNonNullFeedbackSession`?

File: src/main/java/teammates/ui/webapi/action/PublishFeedbackSessionAction.java [hunk 4]
```diff
@@ -24,10 +24,10 @@ public void checkSpecificAccessControl() {
         String courseId = getNonNullRequestParamValue(Const.ParamsNames.COURSE_ID);
         String feedbackSessionName = getNonNullRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_NAME);
 
+        FeedbackSessionAttributes feedbackSession = getFeedbackSession(feedbackSessionName, courseId);
         InstructorAttributes instructor = logic.getInstructorForGoogleId(courseId, userInfo.getId());
-        FeedbackSessionAttributes session = logic.getFeedbackSession(feedbackSessionName, courseId);
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/webapi/action/PublishFeedbackSessionAction.java:28]: Unnecessary line move

File: src/main/java/teammates/ui/webapi/action/UnpublishFeedbackSessionAction.java [hunk 5]
```diff
@@ -24,10 +24,10 @@ public void checkSpecificAccessControl() {
         String courseId = getNonNullRequestParamValue(Const.ParamsNames.COURSE_ID);
         String feedbackSessionName = getNonNullRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_NAME);
 
+        FeedbackSessionAttributes feedbackSession = getFeedbackSession(feedbackSessionName, courseId);
         InstructorAttributes instructor = logic.getInstructorForGoogleId(courseId, userInfo.getId());
-        FeedbackSessionAttributes session = logic.getFeedbackSession(feedbackSessionName, courseId);
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/webapi/action/UnpublishFeedbackSessionAction.java:28]: Similarly here

**Review by wkurniawan07 — APPROVED** [review]
LGTM

**Pull request merged**